qep-13-efficient-snapping-and-geometry-queries.rst

```diff
@@ -0,0 +1,234 @@
+- R-tree for areas stores bounding boxes of individual polygons and their GEOS geometry
+
+
+Examples
+--------
+
+1. snap to a point according to project's snapping settings::
+
+    m = iface.mapCanvas().snappingUtils().snapToMap(QgsPoint(11,22))
+  
+    if not m.isValid():
+      print "no match!"
+      return
+
+    print "match: ", m.point(), m.distance(), m.layer(), m.featureId()
+
+
+2. do queries on a particular layer::
+
+      # get the point locator: uses map units
+      locator = iface.mapCanvas().snappingUtils().locatorForLayer(layer)
+      
+      # find the nearest vertex and edge (no maximum tolerance)
+      mV = locator.nearestVertex(QgsPoint(11,22))
+      mE = locator.nearestEdge(QgsPoint(11,22))
+      
+      # find 5 nearest vertices and edges (no maximum tolerance)
+      lstV = locator.nearestVertices(QgsPoint(11,22), 5)
+      lstE = locator.nearestEdges(QgsPoint(11,22), 5)
+      
+      # find the nearest vertex within tolerance
+      lstV = locator.verticesInTolerance(QgsPoint(33,44), 10)
+      lstE = locator.edgesInTolerance(QgsPoint(33,44), 10)
+      
+      # find out in which polygons the point is located
+      for m in locator.pointInPolygon(QgsPoint(33,44)):
+        print "pt in polygon: ", m.featureId()
+
+
+3. custom point locator - useful for analytic tools working without map canvas::
+
+      locator = QgsPointLocator(layer)
+      
+      m = locator.nearestVertex(QgsPoint(1,1))
+
+4. custom snapping utils - useful for analytic tools working without map canvas::
+
+      utils = QgsSnappingUtils()
+      utils.setMapSettings(settings)
+      utils.setSnapToMapMode(QgsSnappingUtils.SnapAdvanced)
+      cfg1 = QgsSnappingUtils.LayerConfig(layer1, QgsPointLocator.Vertex, 0.1, QgsTolerance.MapUnits)
+      cfg2 = QgsSnappingUtils.LayerConfig(layer2, QgsPointLocator.Edge, 0.2, QgsTolerance.MapUnits)
+      utils.setLayers([cfg1, cfg2])
+      
+      m = utils.snapToMap(QgsPoint(11,22))
+
+
+GUI Changes
+-----------
+
+The snapping settings dialog will be updated to support the new mode "snap to current layer".
+The new mode will be the default.
+The user will be able to choose snapping type and tolerance.
+The existing snapping settings (with a table listing individual layers) will be marked as "advanced" mode.
+
+
+
+Performance Implications
+------------------------
+
+It is expected that snapping performance will be sped up significantly.
+From some quick tests, the current snapping took ~30ms to find the closest vertex, while with new implementation
+needed only less than 1ms. This is because current snapping needs to hit data provider (if the layer is not in editing mode).
+
+There is some cost in initial indexing in ``QgsPointLocator``. For a layer with ~50K points this took about 100ms.
+This is just one-time cost to build the index from data provider's features when queries are first needed.
+No extra cost when map is rendered.
+
```

Comment:
In the current implementation yes the whole layer is cached in memory (not automatically - only when first needed) - though it should be fairly easy to use spatial index library's disk storage instead of memory storage for big layers.